### PR TITLE
HARU-39 - Login signup zod validation

### PR DESCRIPTION
**Pull request by emilielydiat**
- Implement React Hook Form and Zod validation on login page and signup page
- Update actions.ts to manage form submission

### CODE REVIEW

File: app/login/page.tsx [hunk 1]
```diff
@@ -1,21 +1,47 @@
...
 
-  const signInAction = async (formData: FormData) => {
-    'use server';
-    await signIn('credentials', {
-      redirectTo: '/',
-      email: formData.get('email') as string,
-      password: formData.get('password') as string,
-    });
-  };
+  const schema = z.object({
```
Comment (patrickchin) [inline app/login/page.tsx:17]: Can be moved to types.ts so it can be accessed by the server

File: app/login/page.tsx [hunk 2]
```diff
@@ -1,21 +1,47 @@
...
+  const onSubmit: SubmitHandler<FormFields> = async (data) => {
+
+    try {
+      await signInFromLogin({
+        redirectTo: '/',
+        email: data.email,
+        password: data.password,
+      });
+    } catch (error) {
+      console.error("Error:", error);
```
Comment (patrickchin) [inline app/login/page.tsx:42]: This results in "Error: error: blah"
Can set form error state instead

File: app/login/page.tsx [hunk 3]
```diff
@@ -26,24 +52,30 @@ export default function Login() {
...
 
           <div>
             <Label htmlFor="email" className="text-xs uppercase">
               Email Address
-            </Label>
-            <Input name="email" type="email" placeholder="user@acme.com"
+              <Input {...register("email")} type="email" placeholder="user@acme.com"
               autoComplete="email" required className="text-sm" />
+            </Label>
+            {errors.email && (
```
Comment (patrickchin) [inline app/login/page.tsx:63]: Use shadcn Form and FormErrorMessage for these error states

File: app/register/page.tsx [hunk 4]
```diff
@@ -1,34 +1,55 @@
...
-    // TODO do confirm password checking locally with react-hook-forms
-    if (confirmPassword !== password) {
-      return 'Passwords do not match';
-    }
+  const schema = z.object({
+    name: z.string().trim().min(1, {message: "Name is required"}),
+    phone: z.string().min(5, {message: "Phone must contain at least 5 characters"}).max(32, {message: "Phone cannot contain more than 32 characters"}).regex(/^\+?[0-9\s-]+$/, {message: "Phone contains invalid characters"}).optional().or(z.literal("")),
+    email: z.string().trim().min(1, {message: "Email is required"}).email(),
+    password: z.string().min(8, {message: "Password must contain at least 8 characters"}),
+    confirmPassword: z.string().min(8, {message: "Password must contain at least 8 characters"}),
```
Comment (patrickchin) [inline app/register/page.tsx:21]: I think this message is not required

File: app/register/page.tsx [hunk 5]
```diff
@@ -1,34 +1,55 @@
...
 
-    if (user.length > 0) {
-      // TODO: Handle errors with useFormStatus
-      return 'User already exists';
-    } else {
-      await createUser(name, phone, email, password);
+  const onSubmit: SubmitHandler<FormFields> = async (data) => {
+    
+    try {
+      let user = await getUserFromRegister(data.email);
```
Comment (patrickchin) [inline app/register/page.tsx:41]: We don't want to do this validation client side, this way someone can call createUserFromRegister(data) directly and change an existing users email

Fix you one!

File: lib/actions.ts [hunk 6]
```diff
@@ -1,15 +1,28 @@
...
+import { signIn } from '@/lib/auth';
+import { auth } from './auth';
+import * as blob  from '@vercel/blob';
+import { NewProjectFormSchema } from './types';
+import { redirect } from 'next/navigation';
+import { defaulTaskSpecs } from './tasks';
 
 const VERCEL_BLOB_FAKE_FILES = true;
 
+export async function createUserFromRegister(data: any) {
```
Comment (patrickchin) [inline lib/actions.ts:14]: A defined type rather than any is much better here

**Review by patrickchin — CHANGES_REQUESTED** [review]
Fix the hacking vulnerability and it should be fine to merge, we can come back to the other comments later. 

Also run the prettier command on the files you've changed (not all the files)

File: app/register/page.tsx (hunk 5 above)
Comment (patrickchin) [reply app/register/page.tsx:41]: Checks should always be done serverside in actions.ts. client side checks are for UX, making it easier to use. Server side checks are for security and business logic

File: app/register/page.tsx [hunk 7]
```diff
@@ -40,48 +61,61 @@ export default function Page() {
...
+            {errors.name && (
+              <p className="text-sm font-medium text-destructive">{errors.name.message}</p>
+            )}
           </div>
 
           <div>
             <Label htmlFor="phone" className="text-xs uppercase">
               Phone Number
+              <Input {...register("phone")} className="text-sm"
+              type="tel" placeholder="+254" />
```
Comment (patrickchin) [inline app/register/page.tsx:80]: Eventually would like a country selector

File: app/login/page.tsx [hunk 8]
```diff
@@ -1,21 +1,47 @@
...
-  const signInAction = async (formData: FormData) => {
-    'use server';
-    await signIn('credentials', {
-      redirectTo: '/',
-      email: formData.get('email') as string,
-      password: formData.get('password') as string,
-    });
-  };
+  const schema = z.object({
+    email: z.string().trim().min(1, {message: "Email is required"}).email(),
```
Comment (patrickchin) [inline app/login/page.tsx:18]: Looking at this again, I'm not sure if the trim actually changes the data being sent to the database or if it just does the trim for the validation. Could you double check this? It would be odd if there was an email in the database that started with a space or something

File: app/register/page.tsx (hunk 4 above)
Comment (emilielydiat) [reply app/register/page.tsx]: Makes sense, updated

File: app/register/page.tsx (hunk 5 above)
Comment (emilielydiat) [reply app/register/page.tsx]: Fixed - let me know if it's alright now?

File: app/login/page.tsx (hunk 8 above)
Comment (emilielydiat) [reply app/login/page.tsx]: Checked and trim doesn't change the data being sent to the database - removed from both register & login pages

File: lib/actions.ts (hunk 6 above)
Comment (emilielydiat) [reply lib/actions.ts]: Updated!

File: lib/actions.ts [hunk 9]
```diff
@@ -2,15 +2,39 @@
...
+  try {
+    let user = await db.getUser(data.email);
+
+    if (user.length > 0) {
+      throw new Error("User already exists");
+    }
+
+    await db.createUser(data.name, data.phone ?? "", data.email, data.password);
+  } catch (error) {
+    throw error;
```
Comment (patrickchin) [inline lib/actions.ts:25]: catch throw error is the same as if you didn't have a try catch block

File: lib/actions.ts [hunk 10]
```diff
@@ -2,15 +2,39 @@
...
+  }
+}
+
+export async function signInFromLogin(data: any) {
```
Comment (patrickchin) [inline lib/actions.ts:29]: actually, this function isn't needed, you can call `signIn("credentials",`directly on the front end

File: lib/actions.ts [hunk 11]
```diff
@@ -2,15 +2,39 @@
...
+import { NewProjectFormSchema, RegisterSchemaType } from "./types";
 import { redirect } from "next/navigation";
 import { defaulTaskSpecs } from "./tasks";
 import assert from "assert";
 
 const VERCEL_BLOB_FAKE_FILES = true;
 
+export async function registerUser(data: RegisterSchemaType) {
+  try {
+    let user = await db.getUser(data.email);
```
Comment (patrickchin) [inline lib/actions.ts:17]: Another thing to consider is synchronicity
it's possible for two people to try and register with the same email at the same time. In this case two registerUser functions would be run at the same time (I assume this is how the nextjs backend works) and both of those function instances would return no users here and both would call createuser.

However , this is not such a big deal as the email schema labels email as unique, and database edits like this happen one by one, and not in parallel.
https://github.com/patrickchin/haru2-vercel/blob/manage/drizzle/schema.ts#L22

It is good practice to do things "atomically", using a transaction. I recently added one here on this branch:
https://github.com/patrickchin/haru2-vercel/blob/manage/lib/db.ts#L157
In this case, i can't delete the team until all team members are deleted. But to make sure no team members are added between deleting all the members and deleting the team, they have to be done in a transaction so no other changes can be made to the database in the meantime. Or if there are, you won't get partial edits of the database and it will roll back all of the changes.

TBH this is a whole lot of yapping, and I just realised you don't actually need to do a transaction or even do this check, the .unique() already does that check for you

File: app/register/page.tsx [hunk 12]
```diff
@@ -1,87 +1,139 @@
...
-      return 'User already exists';
-    } else {
-      await createUser(name, phone, email, password);
-      redirect('/login');
+  const onSubmit: SubmitHandler<RegisterSchemaType> = async (data) => {
+    try {
+      await registerUser(data);
+      redirect("/login");
+    } catch (error) {
+      console.error("Error:", error);
```
Comment (patrickchin) [inline app/register/page.tsx:30]: We can leave this for another PR, but I think we can add better error handling here, and show an error state in the UI

File: app/login/page.tsx [hunk 13]
```diff
@@ -1,49 +1,93 @@
...
-      redirectTo: '/',
-      email: formData.get('email') as string,
-      password: formData.get('password') as string,
+  const onSubmit: SubmitHandler<FormFields> = async (data) => {
+    const ret = await signInFromLogin({
+      redirectTo: "/",
+      email: data.email,
+      password: data.password,
     });
+    console.log(ret);
```
Comment (patrickchin) [inline app/login/page.tsx:37]: same as the register onSubmit, we could add some UI to tell the user that the login failed or succeeded and is being redirected

File: lib/actions.ts (hunk 10 above)
Comment (patrickchin) [reply lib/actions.ts:29]: Not 100% if this is a good idea though, if you could look up some other examples. does nextauth validate the data itself? I assume so, but it would be good to double check

**Comment by patrickchin** [comment]
I have disabled tests for now, will merge this now and we can come back to the comments in a future PR
